### ISSUE TITLE
sql: implement json compare

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#43123]
While working on sqlsmith, we discovered that the json type doesn't implement the < and <= (and > and >=) operators. There's a json.Compare method, so I went ahead and just added it to the list, assuming that it had been merely forgotten. I wrote a test to generate some json stuff and compare them, and ran it against postgres. The results differed, and should probably be fixed before we add this feature to postgres.

```
WITH
	t (c)
		AS (
			VALUES
				('1'::JSONB),
				('true'::JSONB),
				('[]'::JSONB),
				('{}'::JSONB)
		)
SELECT
	a.c, b.c, a.c < b.c, a.c <= b.c
FROM
	t AS a, t AS b
ORDER BY
	a.c, b.c;
```

cockroach:
```
   c   |  c   | ?column? | ?column?  
+------+------+----------+----------+
     1 |    1 |  false   |   true    
     1 | []   |   true   |   true    
     1 | true |   true   |   true    
     1 | {}   |   true   |   true    
  []   |    1 |  false   |  false    
  []   | []   |  false   |   true    
  []   | true |  false   |  false    
  []   | {}   |   true   |   true    
  true |    1 |  false   |  false    
  true | []   |   true   |   true    
  true | true |  false   |   true    
  true | {}   |   true   |   true    
  {}   |    1 |  false   |  false    
  {}   | []   |  false   |  false    
  {}   | true |  false   |  false    
  {}   | {}   |  false   |   true  
```

postgres:
```
  c   |  c   | ?column? | ?column? 
------+------+----------+----------
 []   | []   | f        | t
 []   | {}   | t        | t
 []   | 1    | t        | t
 []   | true | t        | t
 {}   | []   | f        | f
 {}   | {}   | f        | t
 {}   | 1    | f        | f
 {}   | true | f        | f
 1    | []   | f        | f
 1    | {}   | t        | t
 1    | 1    | f        | t
 1    | true | t        | t
 true | []   | f        | f
 true | {}   | t        | t
 true | 1    | f        | f
 true | true | f        | t
```

Although the orders differ here (which is weird?), the output also differs. Take the `1, []` case, where postgres says `false, false` and cockroach says `true, true`.

Jira issue: CRDB-5297

**Comment by mjibson** [comment]
@justinj do you have memories about why this is? For example https://github.com/cockroachdb/cockroach/blob/1c54631f56ad3c4854bd8125d89967318ccb2cee/pkg/util/json/json_test.go#L73 says that we know we differ from postgres. Do you recall a technical reason why fixing this would be hard?

**Comment by justinj** [comment]
I don't think there's a technical reason, I didn't do it because I thought it would be tricky/error prone (but possible) and not that helpful, but it's caused problems so if someone has the bandwidth to do it it would probably be worthwhile.

**Comment by knz** [comment]
The pg docs have an entry somewhere that details the json compare algorithm. If we add the operator we should follow that doc to the letter.

**Comment by knz** [comment]
There's also an entry about array compares.

**Comment by michae2** [comment]
[triage] @mgartner to verify that this is now fixed

**Comment by mgartner** [comment]
Comparison of JSON values was added in #99275 and adjusted to match Postgres in #101260, both of which will be included in v23.2.
